Src/Compilers/Core/MSBuildTask/BuildClient.cs

```diff
@@ -0,0 +1,398 @@
+                    // Check for already running processes in case someone came in before us
+                    pipe = TryExistingProcesses(expectedServerExePath, cancellationToken);
+                    if (pipe != null)
+                    {
+                        return TryCompile(pipe, request, cancellationToken);
+                    }
+                    else
+                    {
+                        int processId = TryCreateServerProcess(expectedServerExePath);
+                        if (processId != 0 &&
+                            null != (pipe = TryConnectToProcess(processId,
+                                                                TimeOutMsNewProcess,
+                                                                cancellationToken)))
+                        {
+                            // Let everyone else access our process
+                            mutex.ReleaseMutex();
+
+                            return TryCompile(pipe, request, cancellationToken);
+                        }
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Try to compile using the server. Returns null if a response from the
+        /// server cannot be retrieved.
+        /// </summary>
+        private static async Task<BuildResponse> TryCompile(NamedPipeClientStream pipeStream,
+                                                            BuildRequest request,
+                                                            CancellationToken cancellationToken)
+        {
+            BuildResponse response;
+            using (pipeStream)
+            {
+                // Write the request
+                try
+                {
+                    Log("Begin writing request");
+                    await request.WriteAsync(pipeStream, cancellationToken).ConfigureAwait(false);
+                    Log("End writing request");
+                }
+                catch (Exception e)
+                {
+                    LogException(e, "Error writing build request.");
+                    return null;
+                }
+
+                // Wait for the compilation and a monitor to dectect if the server disconnects
+                var serverCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+                Log("Begin reading response");
+
+                var responseTask = BuildResponse.ReadAsync(pipeStream, serverCts.Token);
+                var monitorTask = CreateMonitorDisconnectTask(pipeStream, serverCts.Token);
+                await Task.WhenAny(responseTask, monitorTask).ConfigureAwait(false);
+
+                Log("End reading response");
+
+                if (responseTask.IsCompleted)
+                {
+                    // await the task to log any exceptions
+                    try
+                    {
+                        response = await responseTask.ConfigureAwait(false);
```

Comment:
It looks like you could directly call the `Result` property, because the task is already completed.